### PR TITLE
Support tags on fixture objects (closes #650)

### LINKED ISSUE #650: Support tags on fixture objects, automatically tagging client tests

### PR DESCRIPTION
**Pull request by maormarcus**

### CODE REVIEW

File: slash/core/function_test.py [hunk 1]
```diff
@@ -16,7 +16,11 @@ def __init__(self, function, fixture_store, fixture_namespace, variation):
         self._func = function
 
     def get_tags(self):
-        return get_tags(self._func)
+        test_tags = get_tags(self._func)
+        if not nofixtures.is_marked(self._func):
+            for fixture in list(self.get_required_fixture_objects()):
+                test_tags += fixture.get_tags(self._fixture_store)
```
Comment (vmalloc) [inline slash/core/function_test.py]: Can’t this be done with a single `extend` call and a generator comprehension?

File: slash/core/test.py [hunk 2]
```diff
@@ -75,7 +75,11 @@ def get_test_function(self):
         return getattr(self, self._test_method_name)
 
     def get_tags(self):
-        return get_tags(type(self)) + get_tags(getattr(type(self), self._test_method_name))
+        test_tags = get_tags(type(self)) + get_tags(getattr(type(self), self._test_method_name))
+        if not nofixtures.is_marked(self.get_test_function()):
+            for fixture in list(self.get_required_fixture_objects()):
+                test_tags += fixture.get_tags(self._fixture_store)
```
Comment (vmalloc) [inline slash/core/test.py]: This is duplicated code...

File: tests/test_tagging.py [hunk 3]
```diff
@@ -94,6 +94,33 @@ def test_something():
...
+        @slash.tag('test-tag')
+        def test_something(fixture2):  # pylint: disable=unused-argument
+            pass
+
+    path = suite_builder.build().path
+
+    with slash.Session():
+        test = Loader().get_runnables(path)[0]
+        for tag_name in ['fixture1-tag', 'fixture2-tag', 'test-tag']:
+            assert tag_name in test.get_tags()
```
Comment (vmalloc) [inline tests/test_tagging.py]: This doesn’t test the absence of the tags on tests that shouldn’t have it. You need to add a second test depending on only one fixture, and a third depending on none. 

I would also add a case of both test and fixture containing the same tag...

File: slash/core/fixtures/fixture.py [hunk 4]
```diff
@@ -26,6 +27,13 @@ def __init__(self, store, fixture_func):
         self.scope = self.info.scope
         self.namespace = Namespace(store, store.get_current_namespace())
 
+    def get_tags(self, store):
+        fixture_tags = get_tags(self.fixture_func)
+        required_fixtures = store.get_required_fixture_objects(self.fixture_func, self.namespace)
+        while required_fixtures:
+            fixture_tags += (required_fixtures.pop().get_tags(store))
```
Comment (vmalloc) [inline slash/core/fixtures/fixture.py]: Unnecessary parentheses?

Comment (vmalloc) [reply slash/core/fixtures/fixture.py]: Also I still don’t understand why this isn’t a generator/list comprehension

File: slash/core/fixtures/fixture.py [hunk 5]
```diff
@@ -26,6 +27,13 @@ def __init__(self, store, fixture_func):
         self.scope = self.info.scope
         self.namespace = Namespace(store, store.get_current_namespace())
 
+    def get_tags(self, store):
+        fixture_tags = get_tags(self.fixture_func)
```
Comment (vmalloc) [inline slash/core/fixtures/fixture.py]: Should be:
```
fixture_tags = get_tags(...).copy()
fixture_tags.update(
   fixture.get_tags(store) for fixture in store.get_required_fixture_objects(...))
```

File: slash/core/fixtures/fixture.py [hunk 6]
```diff
@@ -26,6 +27,13 @@ def __init__(self, store, fixture_func):
         self.scope = self.info.scope
         self.namespace = Namespace(store, store.get_current_namespace())
 
+    def get_tags(self, store):
+        current_fixture_tags = get_tags(self.fixture_func)
+        returned = current_fixture_tags.copy() if current_fixture_tags is not NO_TAGS else Tags()
+        required_fixtures_tags = [fixture.get_tags(store) for fixture in store.get_required_fixture_objects(self.fixture_func, self.namespace)]
+        any(returned.update(tags) for tags in required_fixtures_tags)
```
Comment (vmalloc) [inline slash/core/fixtures/fixture.py]: change to `for` loop

**Comment by vmalloc** [comment]
Thanks!
